core/src/main/java/org/elasticsearch/index/query/IndicesQueryParser.java

```diff
@@ -58,40 +44,34 @@ public IndicesQueryParser(@Nullable ClusterService clusterService, IndexNameExpr
     }
 
     @Override
-    public Query parse(QueryShardContext context) throws IOException, QueryParsingException {
-        QueryParseContext parseContext = context.parseContext();
+    public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOException, QueryParsingException {
         XContentParser parser = parseContext.parser();
 
-        Query noMatchQuery = null;
-        boolean queryFound = false;
-        boolean indicesFound = false;
-        boolean currentIndexMatchesIndices = false;
+        QueryBuilder innerQuery = null;
+        Collection<String> indices = new ArrayList<>();
+        QueryBuilder noMatchQuery = QueryBuilders.matchAllQuery();
+
         String queryName = null;
         float boost = AbstractQueryBuilder.DEFAULT_BOOST;
 
         String currentFieldName = null;
         XContentParser.Token token;
-        XContentStructure.InnerQuery innerQuery = null;
-        XContentStructure.InnerQuery innerNoMatchQuery = null;
         while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
             if (token == XContentParser.Token.FIELD_NAME) {
                 currentFieldName = parser.currentName();
             } else if (token == XContentParser.Token.START_OBJECT) {
                 if (parseContext.parseFieldMatcher().match(currentFieldName, QUERY_FIELD)) {
-                    innerQuery = new XContentStructure.InnerQuery(parseContext, null);
-                    queryFound = true;
+                    innerQuery = parseContext.parseInnerQueryBuilder();
                 } else if (parseContext.parseFieldMatcher().match(currentFieldName, NO_MATCH_QUERY)) {
-                    innerNoMatchQuery = new XContentStructure.InnerQuery(parseContext, null);
+                    noMatchQuery = parseContext.parseInnerQueryBuilder();
                 } else {
                     throw new QueryParsingException(parseContext, "[indices] query does not support [" + currentFieldName + "]");
                 }
             } else if (token == XContentParser.Token.START_ARRAY) {
                 if ("indices".equals(currentFieldName)) {
-                    if (indicesFound) {
+                    if (!indices.isEmpty()) {
```

Comment:
Ok the reason why I asked you to change this is that the exclamation mark can easily disappear or trick people. In general we prefer to be explicit and add `== false`, that's what we do in master and we should do the same here. Can you please make this change?